src/recorder.ts

```diff
@@ -13,13 +15,26 @@ import { getTime } from "./util/getTime";
 
 export let recording: Recording;
 
+const recordCallPackageStack: (Package | undefined)[] = [];
+
+function shallowModeSkip(funInfo: FunctionInfo, isLibrary: boolean) {
+  const pkg = config.getPackage(funInfo.location?.path, isLibrary);
+  // If we have the same package as a parent we should skip recording
+  // in shallow mode
+  return pkg?.shallow && recordCallPackageStack.includes(pkg);
+}
+
 export function record<This, Return>(
   this: This,
   fun: (this: This, ...args: unknown[]) => Return,
   args: unknown[],
   funInfo: FunctionInfo,
+  isLibrary = false,
 ): Return {
-  if (!recording.running || !shouldRecord()) return fun.apply(this, args);
+  if (!recording.running || !shouldRecord() || shallowModeSkip(funInfo, isLibrary))
+    return fun.apply(this, args);
+
+  recordCallPackageStack.push(config.getPackage(funInfo.location?.path, isLibrary));
```

Comment:
I'm caching it inside `config`. Do you think it's worth caching it here again as well?

https://github.com/getappmap/appmap-node/blob/1407a2aefc67ebd21fe799909bd4d0f353800388/src/config.ts#L80-L86

Reply:
I think it makes sense to cache per-funInfo. Imagine if the function is in a hot loop... much faster to lookup by pointer than building and hashing a key.

Reply:
Now:
1. Packages are cached with `funInfo` as the key. (`config.getPackage` stays the same, since we are using it elsewhere, too).
2. Local `getPackage` is called only once if needed.
3. It isn't called when we short circuit with no active recordings or `!souldRecord()`.
4. `pkg` is assigned inline in order to prevent breaking short circuit checks and prevent repeating `return fun.apply(this, args)`.

https://github.com/getappmap/appmap-node/blob/153af06ce9133992e8adb11e65ca716ac4abf923/src/recorder.ts#L60-L66